### PR TITLE
Reapply #3878 + bugfix hit rate tracking

### PR DESCRIPTION
**Pull request by aerosol**
### Changes

This PR re-applies [#3878](https://github.com/plausible/analytics/pull/3878) with one extra commit on top of it: https://github.com/plausible/analytics/pull/3891/commits/6019ebe14681497d2f64f693df423cf5f2201d4f

Previous approach wrongly assumed we can detect if an item was warm or cold on `get_or_store`.
Ideally the telemetry (or any functionally equivalent approach) would have been implemented in con_cache - [I guess we'll find out](https://github.com/sasa1977/con_cache/issues/75). Luckily for us, `con_cache` exposes low-level `Operations` interface for which I'm very grateful :)

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [x] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [ ] The UI has been tested both in dark and light mode
- [x] This PR does not change the UI

### CODE REVIEW

File: mix.lock [hunk 1]
```diff
@@ -8,7 +8,6 @@
...
   "bunt": {:hex, :bunt, "1.0.0", "081c2c665f086849e6d57900292b3a161727ab40431219529f13c4ddcf3e7a44", [:mix], [], "hexpm", "dc5f86aa08a5f6fa6b8096f0735c4e76d54ae5c9fa2c143e5a1fc7c1cd9bb6b5"},
   "bypass": {:hex, :bypass, "2.1.0", "909782781bf8e20ee86a9cabde36b259d44af8b9f38756173e8f5e2e1fabb9b1", [:mix], [{:plug, "~> 1.7", [hex: :plug, repo: "hexpm", optional: false]}, {:plug_cowboy, "~> 2.0", [hex: :plug_cowboy, repo: "hexpm", optional: false]}, {:ranch, "~> 1.3", [hex: :ranch, repo: "hexpm", optional: false]}], "hexpm", "d9b5df8fa5b7a6efa08384e9bbecfe4ce61c77d28a4282f79e02f1ef78d96b80"},
-  "cachex": {:hex, :cachex, "3.6.0", "14a1bfbeee060dd9bec25a5b6f4e4691e3670ebda28c8ba2884b12fe30b36bf8", [:mix], [{:eternal, "~> 1.2", [hex: :eternal, repo: "hexpm", optional: false]}, {:jumper, "~> 1.0", [hex: :jumper, repo: "hexpm", optional: false]}, {:sleeplocks, "~> 1.1", [hex: :sleeplocks, repo: "hexpm", optional: false]}, {:unsafe, "~> 1.0", [hex: :unsafe, repo: "hexpm", optional: false]}], "hexpm", "ebf24e373883bc8e0c8d894a63bbe102ae13d918f790121f5cfe6e485cc8e2e2"},
```
Comment (ruslandoga) [inline mix.lock:11]: `mix deps.clean --unused --unlock` also removes and unlocks the following deps for me:

```
* eternal
* jumper
* sleeplocks
* unsafe
```

File: lib/plausible/cache/stats.ex [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  def record_miss(cache_name) do
+    :telemetry.execute(@telemetry_miss, %{}, %{cache_name: cache_name})
+  end
+
+  def init(nil) do
+    __MODULE__ =
+      :ets.new(__MODULE__, [
+        :public,
+        :named_table,
+        :ordered_set,
```
Comment (ruslandoga) [inline lib/plausible/cache/stats.ex:31]: I wonder why not just `:set`? It doesn't seem to do range queries.

Comment (ruslandoga) [reply lib/plausible/cache/stats.ex:31]: I ran a benchmark and there doesn't seem to be any difference in `ips`

I thought `:set` would be faster...

Comment (ruslandoga) [reply lib/plausible/cache/stats.ex:31]: Seems like the telemetry.execute "dominates" over the ets ops hence the little difference. Either way it's super fast.

File: mix.exs [hunk 3]
```diff
@@ -40,7 +40,9 @@ defmodule Plausible.MixProject do
         :logger,
         :runtime_tools,
         :tls_certificate_check,
-        :opentelemetry_exporter
+        :opentelemetry_exporter,
+        :wx,
+        :observer
```
Comment (ruslandoga) [inline mix.exs:45]: I don't know if it affects anything (like increasing the app size), but would it make sense to not include wx and observer in prod builds?

Comment (aerosol) [reply mix.exs:45]: Ah, completely unrelated change, thanks for spotting

Comment (aerosol) [reply mix.exs]: 56d895114

File: lib/plausible/cache/stats.ex (hunk 2 above)
Comment (aerosol) [reply lib/plausible/cache/stats.ex:31]: Oh cool, thanks for running these.

Re:

> I thought :set would be faster...

From your benchmarks it looks like it is, still? Or do you mean you expected a bigger difference? I guess it doesn't cost anything to change it to `:set`, do you agree?

@zoldar even suggested [atomic counters](https://www.erlang.org/doc/man/counters), but I thought with the necessity of keeping refs, it wasn't worth it. WYT?

File: mix.lock (hunk 1 above)
Comment (aerosol) [reply mix.lock:11]: Great, thanks, f38b4067a

File: lib/plausible/cache/stats.ex (hunk 2 above)
Comment (aerosol) [reply lib/plausible/cache/stats.ex]: 9fb8f2bba

Comment (ruslandoga) [reply lib/plausible/cache/stats.ex]: > Or do you mean you expected a bigger difference?

Yes, I expected the difference to be more noticeable.

> I guess it doesn't cost anything to change it to :set, do you agree?

I agree :)

> [atomic counters](https://www.erlang.org/doc/man/counters) WYT?

In my tests for the rate limiter they were somehow slower than ets.update_counter but the use case there was slightly different.

- https://github.com/ruslandoga/rate_limit, atomics were used in rate_limiter and atomics3

I think Stats are already fast enough :)

**Review by ruslandoga — APPROVED** [review]

**Pull request merged**